### ISSUE TITLE
Terraform v0.13.0-rc1 determines incorrect implicit provider when destroying terraform_remote_data resources

### ISSUE THREAD

**Issue by ZymoticB** [hashicorp/terraform#25719]
### Terraform Version

```
v0.13.0-rc1 / master
```

### Terraform Configuration Files

This can be reproduced by creating a very simple module that outputs the ID of a null resource
```terraform
resource "null_resource" "resource1" {
}

output "null_id" {
  value = null_resource.resource1.id
}
```

applying that, and consuming it as remote data

```terraform
data "terraform_remote_state" "data1" {
  backend = "local"

  config = {
    path = "${path.module}/module1/terraform.tfstate"
  }
}

resource "null_resource" "n1" {
  triggers = {
    id = data.terraform_remote_state.data1.outputs.null_id
  }
}
```
applying this module will get the local state backends ready to reproduce this for reference my test directory looks like
```
.
├── main.tf
├── module1
│   ├── main.tf
│   └── terraform.tfstate
└── terraform.tfstate

1 directory, 4 files
```

By modifying the top `main.tf` to look like
```
resource "null_resource" "n1" {
}
```
we'll no longer be able to plan

### Debug Output

The relevant chunk of the debug output is 
```
2020/07/31 16:02:57 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.MissingProviderTransformer
2020/07/31 16:02:57 [DEBUG] adding implicit provider configuration provider["registry.terraform.io/hashicorp/null"], implied first by null_resource.n1
2020/07/31 16:02:57 [DEBUG] adding implicit provider configuration provider["registry.terraform.io/hashicorp/terraform"], implied first by data.terraform_remote_state.data1
2020/07/31 16:02:57 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.MissingProviderTransformer with new graph:
```

### Expected Behavior
```
null_resource.n1: Refreshing state... [id=5569230757422802683]

------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
-/+ destroy and then create replacement

Terraform will perform the following actions:

  # null_resource.n1 must be replaced
-/+ resource "null_resource" "n1" {
      ~ id       = "5569230757422802683" -> (known after apply)
      - triggers = {
          - "id" = "1546755456574532120"
        } -> null # forces replacement
    }

Plan: 1 to add, 0 to change, 1 to destroy.

------------------------------------------------------------------------

Note: You didn't specify an "-out" parameter to save this plan, so Terraform
can't guarantee that exactly these actions will be performed if
"terraform apply" is subsequently run.
```

### Actual Behavior
```
Error: Provider configuration not present

To work with data.terraform_remote_state.data1 its original provider
configuration at provider["terraform.io/builtin/terraform"] is required, but
it has been removed. This occurs when a provider configuration is removed
while objects created by that provider still exist in the state. Re-add the
provider configuration to destroy data.terraform_remote_state.data1, after
which you can remove the provider configuration again.
```

### Steps to Reproduce
Included in-line with the terraform configs above.

**Comment by danieldreier** [comment]
Thank you for reporting this! I've reproduced this using the ga version of 0.13.0. For the sake of the next person to touch this, I've put the reproduction case in GitHub to minimize copy-paste work: https://github.com/danieldreier/terraform-issue-reproductions/tree/master/25719

**Comment by Siegfriedk** [comment]
How do i actually fix this locally?

If i understand it correctly, terraform expects its builtin terraform provider to be declared and i can't do that as terraform tells me its removed and builtin.

**Comment by apparentlymart** [comment]
Thanks for reporting this, @ZymoticB!

During development of Terraform 0.13 we introduced a special case to make the name `terraform` implicitly to `terraform.io/builtin/terraform` rather than the default of `registry.terraform.io/hashicorp/terraform`, because the latter is a legacy provider intended for Terraform 0.11 users and isn't Terraform 0.13 compatible. However, it seems like we missed updating one place to honor that special case, so it was using the default source address prefix exclusively, even for `terraform`.

I proposed a fix for this over in #25948.

To work around this in the meantime, I think it would work to explicitly remove the remnant resource from the state. In the reproduction case, that would be a command like the following:

```
terraform state rm data.terraform_remote_state.data1
```

With the resource removed from both the configuration and the state, Terraform should no longer feel the need to find a provider configuration for it.

**Comment by apparentlymart** [comment]
Hi @ZymoticB! I'm sorry I didn't see your earlier PR when I was working on this issue. Unfortunately I think my brain has been trained to ignore the UI for inter-issue references and so I skimmed directly down to the comments. :confounded: 

I'm going to merge the PR I submitted just because it has some extra tests, but I'll also raise with the rest of team the idea of finding a clearer way to represent that there's an open PR for a particular issue that someone on the core team didn't originate, so it can appear in our queue of pending PRs rather than in our queue of unexplained issues as this one was.

Thanks for digging into this and submitting the fix, and sorry for the misunderstanding!
